spec/looper_spec.rb

```diff
@@ -0,0 +1,157 @@
+# frozen_string_literal: true
+
+require 'cloud_profiler_agent'
+
+RSpec.describe CloudProfilerAgent::Looper do
+  before do
+    @now = 0.0
+    @sleeps = []
+    @rand = 0.4 # chosen by fair dice roll. guaranteed to be random.
+  end
+
+  subject do
+    described_class.new(
+      clock: -> { @now },
+      sleeper: ->(secs) { @sleeps.push(secs); @now += secs },
+      rander: -> { @rand }
+    )
+  end
+
+  let(:min_time) { subject.min_iteration_time }
+  let(:max_time) { subject.max_iteration_time }
+  let(:backoff) { subject.backoff_factor }
+
+  describe '#run' do
+    it 'runs the block the specified number of times' do
+      runs = []
+      subject.run(3) do
+        runs.push(true)
+      end
+      expect(runs.length).to eq(3)
+    end
+
+    it 'runs the block forever when max_iterations is not given' do
+      # but we don't test this, because how do you test an infinite loop?
```

Comment:
Agreed re: infinite loops. In those scenarios, it seems more important to just test what happens in one tick of the loop vs the loop construct itself. If you really were looking for a way to test you could DI another abstraction that decides when the loop is finished, mock it here, and then test the concrete abstraction logic in another file. TBH, that seems a little more trouble than it's worth here but is an option.